### PR TITLE
Parallelize the Equalization effect

### PR DESCRIPTION
**Pull request by Paul-Licameli**
When you EQ your hour long podcast, your extra cores should not be idle!

(This extends the previously reviewed but still unmerged PR #4016.)

Enjoy this present 🎅 

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/effects/Equalization.cpp [hunk 1]
```diff
@@ -449,11 +459,16 @@ struct EqualizationTask {
...
    float *lastWindow{ window2.get() };
 
+   const WaveTrack &input;
+   Poller &poller;
+
    // create a NEW WaveTrack to hold all of the output,
    // including 'tails' each end
    std::shared_ptr<WaveTrack> output;
 
+   size_t M;
```
Comment (crsib) [inline src/effects/Equalization.cpp:503]: I understand that you are restructuring the old code, but I think that a comment would be very welcome here

Comment (crsib) [reply src/effects/Equalization.cpp:503]: I meant a comment in the source code, not on GitHub

**Review by crsib — APPROVED** [review]
I was quite surprised to observe, that UI thread is still used for the processing. 

No need to change it now, and it doesn't really the performance (I've checked). 

However, as a rule of thumb, we should not do any lengthy operations blocking UI thread.

**Comment by Paul-Licameli** [comment]
> I was quite surprised to observe, that UI thread is still used for the processing.
> 
> No need to change it now, and it doesn't really the performance (I've checked).
> 
> However, as a rule of thumb, we should not do any lengthy operations blocking UI thread.

In the sequential version, the UI thread does everything.

In a parallel version, the UI thread still needs to poll the progress dialog for user cancellation.

If it has no other work to do, then it would need to sleep in a loop to avoid busy-waiting.  Or can you think of an alternative?

If the UI thread then sleeps to let other threads crunch numbers -- then to get polling of the button, wouldn't we need the UI thread no only to sleep and wake up, but also evict some other thread from a processor, only to test the button, sleep again, and impose another context switch to resume the worker?

For these reasons I thought it might actually be simpler and marginally better performance to leave it as I did it.  Correct me if I misunderstand.

**Comment by Paul-Licameli** [comment]
I will soon rebase and force-push this onto more recent master.

I will add a commit with a comment explaining "M".

I must resolve some conflicts in WaveClip and Sequence at the commit
"Growing WaveTrack can retain a minimum buffer in memory until Flush()"

which resulted from the other project about dithering.

**Comment by Paul-Licameli** [comment]
That is now done.  Nothing else changed.

But please notice, the last two commits were only unsuccessful guesses I made to try to get a better speedup for @petersampsonaudacity .

The commit using a splin lock is probably a good idea anyway, but I'm not confident the last one should be included.

Still an unresolved is what more intelligent choice of number of threads we should use.

**Comment by dozzzzer** [comment]
Testing this PR, I've discovered some inaccuracies in calculations, which require further investigation.

In order to observe those, you'll need to do the following:

1. Generate a 30s  white noise clip and save it
2. Using any build of Audacity that doesn't have the parallelization EQ changes, import the noise clip
3. Apply **Graphic EQ** with certain parameters to it
4. Save the project as project A
5. Using the build from this PR (https://github.com/audacity/audacity/pull/4141), import the noise clip 
6. Save the current project as project B
7. Repeat step 2
8. Open up project A and copy the audio clip from it
9. Paste it into project B
10. Now that you have two clips in project B, select one of them and apply "Invert" to it
11. Cmd + A > Mix and Render
12. Switch the scale from Linear to dB
13. Zoom in the resulting audio clip at the very end of it

As a result, you'll get smth like this:

image

The portion of the clip of around 3800 samples long, which will amount to smth around -74 dB. 
The rest of the track, except this small area, shows **-INF dB** meaning that the two clips - the one produced by the application of the sequential EQ effect and the one made by the new parallelized EQ - are perfectly identical and cancel each other when one of them is inverted.

**Comment by petersampsonaudacity** [comment]
@dozzzzer  

Are you exporting WAV or MP3?

If MP3 you would expect that as all MP3 exports pad a little at the beginning of the file.

**Comment by petersampsonaudacity** [comment]
Testing on W10 with 3.2.4 versus @Paul-Licameli  Paul's latest branch build for this: audacity-win-3.3.0-alpha-20230207+1ea47ed-x64

Test project is 2 0m 44s of stereo audio recorded from a webcast _(i.e. a real-life typical large data-set for me)_

1)  Exporting as **MP3 256kbps constant Bitrate Stereo** using the Audacity command **File>Export>Export as MP3**
I get a slight speed improvement:
2.4.3    2min:06sec
Paul's   1min:52sec

2) Exporting as **WAV 16-bit PCM** I get a slightly better improvement
2.4.3    1min:03sec
Paul's   0min:35sec

3) The two exported MP3s are identical
    tested by importing both, iinverting one and mixing - result is a pure flat line

**Comment by crsib** [comment]
> Are you exporting WAV or MP3?

I think @dozzzzer saved the project, no export was involved. Even exporting to WAV is a potentially destructive operation. 

With your second comment I feel that you are commenting in the wrong thread - this PR is about EQ effect, it is not related to Paul's latest efforts on exporting performance.

**Comment by petersampsonaudacity** [comment]
>With your second comment I feel that you are commenting in the wrong thread - this PR is about EQ effect, it is not related to Paul's latest efforts on exporting performance.

@crsib   OOPS, sorry Dmitry  _<blushes deeply>_

I have transferred my export tests that Paul requested to the correct PR thread

**Comment by Paul-Licameli** [comment]
I spoke with @dozzzzer about this.

I recommended going to Interface preferences and choosing the lowest value for cutoff of the dB scale, which made this inaudible anomaly at the end visible in waveform display.

I don't know why it is there and I should figure it out, but it is reassuring that all else in the long track was exactly zero.

@petersampsonaudacity as I also pointed out to @dozzzzer -- for these testing purposes, there should not be any lossy export, or saves to sample format other than float.  Otherwise quantization and dither noise would obscure what is happening.

File: src/effects/Equalization.cpp [hunk 2]
```diff
@@ -411,40 +416,121 @@ bool EffectEqualization::TransferDataToWindow(const EffectSettings &settings)
...
 namespace {
-struct EqualizationTask {
-   EqualizationTask( size_t M, size_t idealBlockLen, WaveTrack &t )
-      : buffer{ idealBlockLen }
+//! Function takes an increment of samples processed, returns true to continue
+using Poller = std::function<bool(sampleCount)>;
+
+class EffectTask {
+public:
+   virtual ~EffectTask();
```
Comment (saintmatthieu) [inline src/effects/Equalization.cpp:424]: Why not directly here `virtual ~EffectTask() = default;` ?

Comment (Paul-Licameli) [reply src/effects/Equalization.cpp:424]: Virtual functions need to have non-inline bodies compiled somewhere.  The compiler might generate them even if you make them all inline, but then it has to eliminate duplicates at link time.

If you make at least one virtual function non-inline, the compiler can follow the rule that it generates the rest of the functions only in the same translation unit and avoid the extra work of removing the duplicates.  So I do this with destructors to be a little nicer to the compiler.

Though sometimes, a template is generating the class, and then you often must inline them anyway and rely on the elimination.

**Review by saintmatthieu — APPROVED** [review]
I really concentrated on the Equalization itself. Did a few tests, LGTM, to the exception of a few thingies I spot. Please consider addressing before merging.

**Comment by dozzzzer** [comment]
Testing this PR now shows the issue remains present. I'm consistently getting about the same result as previously:

image

**Pull request closed without merging**